### PR TITLE
Basic rust indexing support.

### PR DESCRIPTION
**Pull request by emilio**
Don't run it yet by default, since I haven't tested it on mozilla-central.

### CODE REVIEW

File: tools/src/bin/rust-indexer.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        analyze_file(&name, defs, &analysis, src_dir, output_dir);
+    }
+}
+
+fn main() {
+    let matches = app_from_crate!()
+        .args_from_usage(
+            "<src>    'Points to the source root'
+             <input>  'Points to the deps/save-analysis directory'
+             <output> 'Points to the directory where searchfox metadata should'",
```
[the remark this answers is not in the document]
Comment (emilio) [reply tools/src/bin/rust-indexer.rs]: nice catch :)

File: tools/src/bin/rust-indexer.rs [hunk 2]
```diff
@@ -0,0 +1,235 @@
+#[macro_use]
+extern crate clap;
+extern crate rls_analysis as analysis;
```
[the remark this answers is not in the document]
Comment (emilio) [reply tools/src/bin/rust-indexer.rs]: Sure thing, done.

File: tools/src/bin/rust-indexer.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+fn span_to_string(span: &data::SpanData) -> String {
+    // Rust spans are multi-line... So we just set the start column if it spans
+    // multiple rows, searchfox has fallback code to handle this.
+    if span.line_start != span.line_end {
+        return format!("{}:{}", span.line_start.0, span.column_start.0 - 1);
+    }
+    if span.column_start == span.column_end {
+        return format!("{}:{}", span.line_start.0, span.column_start.0 - 1);
+    }
+    let len = span.column_end.0 - span.column_end.0;
```
[the remark this answers is not in the document]
Comment (emilio) [reply tools/src/bin/rust-indexer.rs]: Yup, fixed it.

File: tools/src/bin/rust-indexer.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    format!("{}:{}-{}", span.line_start.0, span.column_start.0 - 1, len)
```
[the remark this answers is not in the document]
Comment (emilio) [reply tools/src/bin/rust-indexer.rs]: Right, that was the fix, thanks! :)

**Comment by emilio** [comment]
I'm trying to run it on central... though it may take a bit :)

**Comment by emilio** [comment]
I added a couple more commits that make this work in mozilla-central. r? @bill-mccloskey

**Comment by bill-mccloskey** [comment]
I tried running this locally on mozilla-central, and I got a segfault while building some Rust code. This was the error.

14:11.94 error: failed to run custom build command for `style v0.0.1 (file:///home/vagrant/mozilla-index/mozilla-central/gecko-dev/servo/components/style)`
14:11.94 process didn't exit successfully: `/home/vagrant/mozilla-index/mozilla-central/objdir/toolkit/library/debug/build/style-348820704fd2a9fe/build-script-build` (signal: 11, SIGSEGV: invalid memory reference)
14:11.94 --- stdout
14:11.94 cargo:rerun-if-changed=build.rs
14:11.94 cargo:out_dir=/home/vagrant/mozilla-index/mozilla-central/objdir/toolkit/library/x86_64-unknown-linux-gnu/debug/build/style-1623c3e1c6bdb176/out
14:11.94 cargo:rerun-if-changed=properties/declaration_block.rs
14:11.94 cargo:rerun-if-changed=properties/properties.mako.rs
14:11.94 cargo:rerun-if-changed=properties/properties.html.mako

Everything else finished successfully, but I didn't get any indexing results for WebRender or Stylo. Some crates did have data, though (libcubeb and a few other things).

Did you see anything like this, Emilio?

**Comment by bill-mccloskey** [comment]
The same thing happens without the -Zsave-analysis option. I wonder if the current Rust nightly is just unable to build Firefox right now? Or maybe it's an issue running inside Vagrant?

**Comment by bill-mccloskey** [comment]
I did have one idea about how to address the issue of how to land this at the same time we want to index C++ code in TaskCluster. It looks like running `mach cargo check gkrust` will generate the save-analysis data. That means we could analyze C++ code in TaskCluster and just run `mach cargo check gkrust` during the Searchfox indexing run, to index only Rust code.

**Comment by emilio** [comment]
The issue you're seeing is https://github.com/rust-lang/rust/issues/46239. I'm landing a workaround in mozilla-central right now.

**Comment by emilio** [comment]
> I did have one idea about how to address the issue of how to land this at the same time we want to index C++ code in TaskCluster. It looks like running mach cargo check gkrust will generate the save-analysis data. That means we could analyze C++ code in TaskCluster and just run mach cargo check gkrust during the Searchfox indexing run, to index only Rust code.

That sounds ok to me, how's the taskcluster stuff happening? I'm happy to try to integrate it that way too.

**Comment by amccreight** [comment]
Do you actually want my review on anything here?

**Comment by emilio** [comment]
I guess billm's review is enough

**Comment by emilio** [comment]
Rebased and verified it still works. The failure on central was fixed long ago, too.

r? @bill-mccloskey

**Comment by staktrace** [comment]
@emilio are there any other changes you want to make here? I looked over the changes (not in much detail though) and they look sane, at least to the point where I'm comfortable merging them and iterating further afterwards. I also created new AMIs with the nightly rust and ran an indexing run with these patches, so the instance up on dev.searchfox.org right now has these changes and they seem to work fine. So if you have no objections I'd like to merge these patches.

**Comment by emilio** [comment]
That'd be great! IIRC @bill-mccloskey mentioned that this was blocked on some kind of infra issue regarding reprovisioning, but if you managed to get it working I think this is good to go.

**Comment by staktrace** [comment]
I've deployed the reprovisioned servers to indexer-16.04 and web-server-16.04 so I think merging these patches should work for tomorrow's indexing run. If something goes wrong I'll investigate.

**Pull request merged**